src/grpc/stream_infer_handler.cc

```diff
@@ -540,57 +544,104 @@ ModelStreamInferHandler::StreamInferResponseComplete(
   }
 
   auto& response_queue = state->response_queue_;
+  // Get request ID from request object to simplify case of null iresponse
+  std::string request_id = state->request_.id();
+  if (request_id.empty()) {
+    request_id = "<id_unknown>";
+  }
 
-  if (iresponse != nullptr) {
-    auto response = response_queue->GetResponseAt(response_index);
-    if (response == nullptr) {
-      LOG_ERROR << "expected the response allocator to have added the response";
-    }
-
+  inference::ModelStreamInferResponse* response = nullptr;
+  if (iresponse) {
+    // Backend returned a non-null response
     TRITONSERVER_Error* err = nullptr;
-    if (iresponse != nullptr) {
+    response = response_queue->GetResponseAt(response_index);
+    if (response) {
       inference::ModelInferResponse& infer_response =
           *(response->mutable_infer_response());
+      // Validate Triton iresponse and set grpc/protobuf response fields from it
       err = InferResponseCompleteCommon<inference::ModelStreamInferResponse>(
           state->tritonserver_, iresponse, infer_response,
           state->alloc_payload_);
+    } else {
+      LOG_ERROR << "expected the response allocator to have added the response";
     }
 
     if (err != nullptr) {
       ::grpc::Status status;
       GrpcStatusUtil::Create(&status, err);
       response->mutable_infer_response()->Clear();
       response->set_error_message(status.error_message());
-
-      // repopulate the id so that client knows which request failed.
-      const char* id;
-      LOG_TRITONSERVER_ERROR(
-          TRITONSERVER_InferenceResponseId(iresponse, &id),
-          "couldn't retrieve id for failed request");
-      LOG_VERBOSE(1) << "Failed for ID: " << id << std::endl;
-      response->mutable_infer_response()->set_id(id);
+      LOG_VERBOSE(1) << "Failed for ID: " << request_id << std::endl;
     }
 
     TRITONSERVER_ErrorDelete(err);
-
     LOG_TRITONSERVER_ERROR(
         TRITONSERVER_InferenceResponseDelete(iresponse),
         "deleting GRPC inference response");
   }
 
-  state->complete_ = ((flags & TRITONSERVER_RESPONSE_COMPLETE_FINAL) != 0);
-  if (!state->is_decoupled_) {
-    state->step_ = Steps::WRITEREADY;
-    state->context_->WriteResponseIfReady(state);
-  } else {
+  // Decoupled backends can return a null response via
+  // TRITONBACKEND_ResponseFactorySendFlags. By default, these null
+  // "empty" responses are not sent back to the client. Clients can
+  // opt-in to receiving these empty responses via request parameters.
+  // NOTE: The complete flag is the only flag used for this case at this time.
+  bool empty = (!iresponse && state->is_decoupled_ && state->complete_);
+  bool enable_empty = false;
+  if (empty) {
+    // Check request for opt-in parameter
+    const auto& request_params = state->request_.parameters();
```

Comment:
For efficiency, how about adding `state` as an arg to [SetInferenceRequestMetadata](https://github.com/triton-inference-server/server/blob/b9e8125eb6fa5a46c0dd3e7010b2a4294f0c04b9/src/grpc/infer_handler.cc#L217) and setting `state->enable_empty_response_` while request parameters are already being parsed there, rather than parsing twice?

Both `state` and `SetInferenceRequestMetadata` are used by both InferHandler and StreamInferHandler, so the `state->enable_empty_response_` wouldn't having meaning for the non-stream infer handler, but this is also true for `is_decoupled_` today.

Something like this:
```cpp
// Already parsing request parameters here when initializing state
TRITONSERVER_Error*
SetInferenceRequestMetadata(
    TRITONSERVER_InferenceRequest* inference_request,
    const inference::ModelInferRequest& request,
    State* state
)
{
...
  for (auto param : request.parameters()) {
    ...
    } else if (param.first.rfind("triton_", 0) == 0) {
      RETURN_IF_ERR(SetStateFromParameter(state, param));
    }
    ...
}
```
and add a new helper:
```cpp
TRITONSERVER_Error*
SetStateFromParameter(
  State* state,
  const inference::InferParameter& param)
{
  const auto& key = param.first;
  const auto& value = param.second;
  if (key == "triton_enable_empty_response") {
    if (value.parameter_choice_case() !=
          inference::InferParameter::ParameterChoiceCase::kBoolParam) {
      return TRITONSERVER_ErrorNew(
          TRITONSERVER_ERROR_INVALID_ARG,
         "invalid value type for 'triton_enable_empty_response' parameter, expected bool_param.");
    }

    // Set state here for use in response complete callback
    state->enable_empty_response_ = value.bool_param();
  } else {
      return TRITONSERVER_ErrorNew(
          TRITONSERVER_ERROR_INVALID_ARG,
          (std::string(
               "parameter keys starting with 'triton_' are reserved for Triton "
               "usage. Only the following keys starting with 'triton_' are "
               "allowed: ") +
           Join(TRITON_RESERVED_REQUEST_PARAMS, " "))
              .c_str());
  } 

  return nullptr;  // success
}
```

Reply:
It's a bit messy in likely having to move `SetInferenceRequestMetadata` into the InferHandler class, similar to [ForwardHeadersAsParameters](https://github.com/triton-inference-server/server/blob/ececc9fb45efd2eab2c9cc9418f80f55dacae9d7/src/grpc/infer_handler.h#L944-L945), and possibly the implementation into the header as well with the templated state types like [here](https://github.com/triton-inference-server/server/blob/ececc9fb45efd2eab2c9cc9418f80f55dacae9d7/src/grpc/infer_handler.h#L1060-L1087), which I don't love.

@nnshah1 Do you have any alternative ideas on a good spot to move this logic without too much mess?

Reply:
unfortunately I think the best option is a new method of stream_infer_handler - we could put decoupled and ```enable_empty_response``` setting into a new method: SetStateOptions

Basically right at stream_infer_handler 200.

Reply:
unfortunately I think the best option is a new method of stream_infer_handler - we could put decoupled and ```enable_empty_response``` setting into a new method: SetStateOptions

Basically right at stream_infer_handler 200. 

> May be we should steer clear of terms "params_only" and "flags_only" altogether to avoid confusion. How about the following?
> 
> ```
> triton_enable_flags_only_response -> triton_enable_final_response
> triton_flags_only_response -> triton_is_final_response
> ```

Summarizing both comments (allowing responses without tensors without specific indication, and providing specific naming to avoid confusion). We would have the following:

``` request.parameters["triton_enable_empty_final_response"]
    response.parameters["triton_final_response"]
```
Empty / non empty would be up to the client to handle appropriately. I think this is the clearest naming for the scenario at hand - probably if we need to be more general in the future we can plan for it then.

Reply:
unfortunately I think the best option is a new method of stream_infer_handler - we could put decoupled and ```enable_empty_response``` setting into a new method: SetStateOptions

Basically right at stream_infer_handler 200. 

> May be we should steer clear of terms "params_only" and "flags_only" altogether to avoid confusion. How about the following?
> 
> ```
> triton_enable_flags_only_response -> triton_enable_final_response
> triton_flags_only_response -> triton_is_final_response
> ```

Summarizing both of @tanmayv25 comments (allowing responses without tensors without specific indication, and providing specific naming to avoid confusion). We would have the following:

``` request.parameters["triton_enable_empty_final_response"]
    response.parameters["triton_final_response"]
```
Empty / non empty would be up to the client to handle appropriately. I think this is the clearest naming for the scenario at hand - probably if we need to be more general in the future we can plan for it then.

Reply:
> May be we should steer clear of terms "params_only" and "flags_only" altogether to avoid confusion. How about the following?
> 
> ```
> triton_enable_flags_only_response -> triton_enable_final_response
> triton_flags_only_response -> triton_is_final_response
> ```

Summarizing both of @Tcomments (allowing responses without tensors without specific indication, and providing specific naming to avoid confusion). We would have the following:

``` request.parameters["triton_enable_empty_final_response"]
    response.parameters["triton_final_response"]
```
Empty / non empty would be up to the client to handle appropriately. I think this is the clearest naming for the scenario at hand - probably if we need to be more general in the future we can plan for it then.